### ISSUE TITLE
assignment 2

### ISSUE THREAD

**Issue by UMAGAYATHRI74** [UMAGAYATHRI74/data-analyst#2]
Assignment 1
Python code:
https://trinket.io/python3/10acfaf40f00
https://www.online-python.com/Me1iUrdunv
import pandas as pd
import matplotlib.pyplot as plt
import os
files = ['Customers.csv', 'Products.csv', 'Transactions.csv']
for file in files:
if not os.path.exists(file):
print(f"Error: File '{file}' not found. Download it and place it in this directory.")
exit()
customers = pd.read_csv('Customers.csv')
products = pd.read_csv('Products.csv')
transactions = pd.read_csv('Transactions.csv')
customers['SignupDate'] = pd.to_datetime(customers['SignupDate'])
transactions['TransactionDate'] = pd.to_datetime(transactions['TransactionDate'])
products['Price'].fillna(products['Price'].mean(), inplace=True)
transactions.dropna(subset=['CustomerID', 'ProductID'], inplace=True)
transactions = transactions.merge(customers[['CustomerID', 'Region']], on='CustomerID',
how='left')
transactions = transactions.merge(products[['ProductID', 'ProductName', 'Category', 'Price']],
on='ProductID', how='left')
print(transactions.describe())
print(transactions['Region'].value_counts())
print(transactions['Category'].value_counts())
def create_and_show_plot(plot_function, data, *args, **kwargs):
plt.figure(figsize=(10, 6))
plot_function(data, *args, **kwargs)
plt.show()
create_and_show_plot(plt.hist, transactions['TotalValue'], bins=30, color='skyblue',
edgecolor='black', label='Total Value')
plt.title('Distribution of Transaction Total Value')
plt.xlabel('Total Value (USD)')
plt.ylabel('Frequency')
plt.legend()
sales_per_region = transactions.groupby('Region')['TotalValue'].sum().reset_index()
create_and_show_plot(plt.bar, sales_per_region['Region'], sales_per_region['TotalValue'],
color='viridis', label='Total Sales')
plt.title('Total Sales per Region')
plt.xlabel('Region')
plt.ylabel('Total Sales (USD)')
plt.legend()
sales_by_category = transactions.groupby('Category')['TotalValue'].sum().reset_index()
create_and_show_plot(plt.barh, sales_by_category['Category'], sales_by_category['TotalValue'],
color='Blues', label='Total Sales')
plt.title('Total Sales by Product Category')
plt.xlabel('Total Sales (USD)')
plt.ylabel('Product Category')
plt.legend()
avg_order_value = transactions.groupby('Region')['TotalValue'].mean().reset_index()
create_and_show_plot(plt.bar, avg_order_value['Region'], avg_order_value['TotalValue'],
color='Set2', label='Average Order Value')
plt.title('Average Order Value by Region')
plt.xlabel('Region')
plt.ylabel('Average Order Value (USD)')
plt.legend()
transactions_per_customer =
transactions.groupby('CustomerID')['TransactionID'].count().reset_index()
transactions_per_customer.columns = ['CustomerID', 'TransactionCount']
create_and_show_plot(plt.hist, transactions_per_customer['TransactionCount'], bins=30,
color='purple', edgecolor='black', label='Transaction Count')
plt.title('Distribution of Transactions per Customer')
plt.xlabel('Number of Transactions')
plt.ylabel('Frequency')
plt.legend()
insights = [
"1. North America exhibits significantly higher total sales compared to other regions.",
"2. The 'Electronics' category generates the highest revenue.",
"3. Customers in Europe tend to have higher average order values.",
"4. Most customers have a low transaction count (1-5), highlighting potential for increased
customer retention and repeat business.",
"5. Lower-priced products drive high sales volume, suggesting promotions on these items
could further boost sales."
]
print("\nBusiness Insights:")
for insight in insights:
print(insight)
transactions.to_csv('transactions_cleaned.csv', index=False)Assingment-2
https://www.online-python.com/E2NbxPqnYJ
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import StandardScaler
try:
customers = pd.read_csv('Customers.csv')
products = pd.read_csv('Products.csv')
transactions = pd.read_csv('Transactions.csv')
except FileNotFoundError:
print("Error: One or more CSV files not found. Please ensure they are in the same directory.")
exit()
customers['SignupDate'] = pd.to_datetime(customers['SignupDate'])
transactions['TransactionDate'] = pd.to_datetime(transactions['TransactionDate'])
products['Price'].fillna(products['Price'].mean(), inplace=True)
transactions.dropna(subset=['CustomerID', 'ProductID'], inplace=True)
transactions = transactions.merge(customers[['CustomerID', 'Region']], on='CustomerID',
how='left')
transactions = transactions.merge(products[['ProductID', 'ProductName', 'Category', 'Price']],
on='ProductID', how='left')
customer_profiles = transactions.groupby('CustomerID').agg(
total_transactions=('TransactionID', 'count'),
total_spent=('TotalValue', 'sum'),
avg_order_value=('TotalValue', 'mean'),
unique_products=('ProductID', 'nunique'),
electronics_purchases=('Category', lambda x: (x == 'Electronics').sum()),
clothing_purchases=('Category', lambda x: (x == 'Clothing').sum()),
home_purchases=('Category', lambda x: (x == 'Home').sum()),
books_purchases=('Category', lambda x: (x == 'Books').sum()),
region=('Region', 'first')
).reset_index()
customer_profiles = pd.get_dummies(customer_profiles, columns=['region'], prefix='region')
numerical_features = ['total_transactions', 'total_spent', 'avg_order_value', 'unique_products',
'electronics_purchases', 'clothing_purchases', 'home_purchases', 'books_purchases']
scaler = StandardScaler()
customer_profiles[numerical_features] =
scaler.fit_transform(customer_profiles[numerical_features])
def find_lookalikes(customer_id, top_n=3):
try:
target_profile = customer_profiles[customer_profiles['CustomerID'] ==
customer_id].drop('CustomerID', axis=1)
if target_profile.empty:
return []
similarities = cosine_similarity(target_profile, customer_profiles.drop('CustomerID', axis=1))
similar_customer_indices = np.argsort(similarities[0])[::-1][1:top_n+1]
lookalikes = customer_profiles.iloc[similar_customer_indices]
scores = similarities[0][similar_customer_indices]
lookalike_list = []
for i in range(len(lookalikes)):
lookalike_list.append({"customer_id": lookalikes['CustomerID'].iloc[i], "similarity_score":
scores[i]})
return lookalike_list
except KeyError:
return []
lookalike_mapping = {}
for i in range(20):
customer_id = f"C{i+1:04d}"
lookalikes = find_lookalikes(customer_id)
lookalike_mapping[customer_id] = lookalikes
lookalike_df = pd.DataFrame.from_dict(lookalike_mapping, orient='index')
lookalike_df = lookalike_df.reset_index().rename(columns={'index': 'customer_id'})
lookalike_df.to_csv('Lookalike.csv', index=False)
print("Lookalike.csv created successfully.")
customer_id_to_check = "C0005"
lookalikes_for_customer = lookalike_mapping.get(customer_id_to_check)
if lookalikes_for_customer:
print(f"Lookalikes for {customer_id_to_check}:")
for lookalike in lookalikes_for_customer:
print(f"Customer ID: {lookalike['customer_id']}, Score: {lookalike['similarity_score']}")
else:
print(f"No lookalikes found for {customer_id_to_check} or customer not found.")

Assignment 3
https://www.online-python.com/h2dMyEnTLx
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans
from sklearn.metrics import davies_bouldin_score, silhouette_score
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
try:
customers = pd.read_csv('Customers.csv')
products = pd.read_csv('Products.csv')
transactions = pd.read_csv('Transactions.csv')
except FileNotFoundError:
print("Error: One or more CSV files not found. Please ensure they are in the same directory.")
exit()
customers['SignupDate'] = pd.to_datetime(customers['SignupDate'])
transactions['TransactionDate'] = pd.to_datetime(transactions['TransactionDate'])
products['Price'].fillna(products['Price'].mean(), inplace=True)
transactions.dropna(subset=['CustomerID', 'ProductID'], inplace=True)
transactions = transactions.merge(customers[['CustomerID', 'Region']], on='CustomerID',
how='left')
transactions = transactions.merge(products[['ProductID', 'ProductName', 'Category', 'Price']],
on='ProductID', how='left')
customer_profiles = transactions.groupby('CustomerID').agg(
total_transactions=('TransactionID', 'count'),
total_spent=('TotalValue', 'sum'),
avg_order_value=('TotalValue', 'mean'),
unique_products=('ProductID', 'nunique'),
electronics_purchases=('Category', lambda x: (x == 'Electronics').sum()),
clothing_purchases=('Category', lambda x: (x == 'Clothing').sum()),
home_purchases=('Category', lambda x: (x == 'Home').sum()),
books_purchases=('Category', lambda x: (x == 'Books').sum()),
region=('Region', 'first')
).reset_index()
customer_profiles = customer_profiles.merge(customers[['CustomerID', 'Region', 'SignupDate']],
on='CustomerID', how='left')
customer_profiles['days_since_signup'] = (pd.to_datetime('today') -
customer_profiles['SignupDate']).dt.days
customer_profiles = pd.get_dummies(customer_profiles, columns=['region'], prefix='region')
customer_profiles.drop('SignupDate', axis=1, inplace=True)
numerical_features = ['total_transactions', 'total_spent', 'avg_order_value', 'unique_products',
'electronics_purchases', 'clothing_purchases', 'home_purchases', 'books_purchases',
'days_since_signup']
scaler = StandardScaler()
customer_profiles[numerical_features] =
scaler.fit_transform(customer_profiles[numerical_features])
best_db_index = -1
best_silhouette_score = -1
best_n_clusters = 0
best_labels = None
best_cluster_centers = None
for n_clusters in range(2, 11):
kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init=10)
labels = kmeans.fit_predict(customer_profiles.drop('CustomerID', axis=1))
db_index = davies_bouldin_score(customer_profiles.drop('CustomerID', axis=1), labels)
silhouette = silhouette_score(customer_profiles.drop('CustomerID', axis=1), labels)
if db_index < best_db_index or best_db_index == -1: # Lower DB index is better
best_db_index = db_index
best_silhouette_score = silhouette
best_n_clusters = n_clusters
best_labels = labels
best_cluster_centers = kmeans.cluster_centers_
customer_profiles['cluster'] = best_labels
print(f"Best number of clusters: {best_n_clusters}")
print(f"DB Index: {best_db_index}")
print(f"Silhouette Score: {best_silhouette_score}")
pca = PCA(n_components=2)
pca_data = pca.fit_transform(customer_profiles.drop(['CustomerID', 'cluster'], axis=1))
customer_profiles['pca_1'] = pca_data[:, 0]
customer_profiles['pca_2'] = pca_data[:, 1]
plt.figure(figsize=(10, 8))
for cluster in range(best_n_clusters):
cluster_data = customer_profiles[customer_profiles['cluster'] == cluster]
plt.scatter(cluster_data['pca_1'], cluster_data['pca_2'], label=f'Cluster {cluster}')
plt.title('Customer Segmentation')
plt.xlabel('PCA Component 1')
plt.ylabel('PCA Component 2')
plt.legend()
plt.show()
for cluster in range(best_n_clusters):
print(f"\nCluster {cluster} Characteristics:")
cluster_data = customer_profiles[customer_profiles['cluster'] == cluster]
print(cluster_data[numerical_features].describe())
print(cluster_data['region_Europe'].value_counts())
